File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    private void dfs(int x, int y, char[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        grid[x][y] = '0';
        for(int[] neighbor: new int[][]{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}) {
            int dx = x + neighbor[0];
            int dy = y + neighbor[1];
            if(0 <= dx && dx < rows && 0 <= dy && dy < cols && grid[dx][dy] == '1') {
                dfs(dx, dy, grid);
            }
        }
    }
    public int numIslands(char[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        int count = 0;
        for(int x = 0; x < rows; x++) {
            for(int y = 0; y < cols; y++) {
                if(grid[x][y] == '1') {
                    count++;
                    dfs(x, y, grid);
                }
            }
        }
        return count;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(i, j):
            grid[i][j] = "2"
            neighbors = [[0, 1], [1, 0], [0, -1], [-1, 0]]
            for di, dj in neighbors:
                if 0 <= i + di <  m and 0 <= j + dj < n and grid[i + di][j + dj] == "1":
                    dfs(i + di, j + dj)
            return 
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    count += 1
                    dfs(i, j)
        return count